src/cargo/core/compiler/fingerprint.rs

```diff
@@ -825,66 +851,62 @@ impl Fingerprint {
     /// an error, it will never return success.
     fn compare(&self, old: &Fingerprint) -> CargoResult<()> {
         if self.rustc != old.rustc {
-            bail!("rust compiler has changed")
+            bail!(DirtyReason::RustcChanged)
         }
         if self.features != old.features {
-            bail!(
-                "features have changed: previously {}, now {}",
-                old.features,
-                self.features
-            )
+            bail!(DirtyReason::FeaturesChanged {
+                old: old.features.clone(),
+                new: self.features.clone()
+            });
         }
         if self.target != old.target {
-            bail!("target configuration has changed")
+            bail!(DirtyReason::TargetConfigurationChanged)
         }
         if self.path != old.path {
-            bail!("path to the source has changed")
+            bail!(DirtyReason::PathToSourceChanged)
         }
         if self.profile != old.profile {
-            bail!("profile configuration has changed")
+            bail!(DirtyReason::ProfileConfigurationChanged)
         }
         if self.rustflags != old.rustflags {
-            bail!(
-                "RUSTFLAGS has changed: previously {:?}, now {:?}",
-                old.rustflags,
-                self.rustflags
-            )
+            bail!(DirtyReason::RustflagsChanged {
+                old: old.rustflags.clone(),
```

Comment:
I'm not sure I understand your concern, but maybe using `InternedString` instead could help? Or alternatively move the

```rs
        if !self.fs_status.up_to_date() {
            bail!(DirtyReason::FsStatusOutdated(FsStatusOutdated::new(
                self.fs_status.clone()
            )))
        }
```

Check before the others, which would force most of the crates to have the reason of "dependency ... was rebuilt" instead of one of the more potentially specific ones. It corresponds to this variant:

https://github.com/rust-lang/cargo/blob/e08848c974c1ab289d6240da113141d999b0639f/src/cargo/core/compiler/fingerprint.rs#L587-L591

Which is also a lot cheaper to clone.

Reply:
I meant avoid some clones like these inside `DirtyReason`. A little overhead is present even during non-verbose mode. Though to fix it, we might introduce quirky lifetime annotations 😞.

Not really a blocker to me if others feel fine with it.

```patch
diff --git a/src/cargo/core/compiler/fingerprint.rs b/src/cargo/core/compiler/fingerprint.rs
index 474ce169f..1e2545878 100644
--- a/src/cargo/core/compiler/fingerprint.rs
+++ b/src/cargo/core/compiler/fingerprint.rs
@@ -855,8 +855,8 @@ impl Fingerprint {
         }
         if self.features != old.features {
             bail!(DirtyReason::FeaturesChanged {
-                old: old.features.clone(),
-                new: self.features.clone()
+                old: &old.features,
+                new: &self.features,
             });
         }
         if self.target != old.target {
@@ -870,8 +870,8 @@ impl Fingerprint {
         }
         if self.rustflags != old.rustflags {
             bail!(DirtyReason::RustflagsChanged {
-                old: old.rustflags.clone(),
-                new: self.rustflags.clone(),
+                old: &old.rustflags,
+                new: &self.rustflags,
             })
         }
         if self.metadata != old.metadata {
diff --git a/src/cargo/core/compiler/fingerprint/dirty_reason.rs b/src/cargo/core/compiler/fingerprint/dirty_reason.rs
index 025aa8ee1..6f0d9f552 100644
--- a/src/cargo/core/compiler/fingerprint/dirty_reason.rs
+++ b/src/cargo/core/compiler/fingerprint/dirty_reason.rs
@@ -5,47 +5,47 @@ use crate::Config;
 use std::fmt;
 
 #[derive(Debug)]
-pub enum DirtyReason {
+pub enum DirtyReason<'a> {
     RustcChanged,
     FeaturesChanged {
-        old: String,
-        new: String,
+        old: &'a str,
+        new: &'a str,
     },
     TargetConfigurationChanged,
     PathToSourceChanged,
     ProfileConfigurationChanged,
     RustflagsChanged {
-        old: Vec<String>,
-        new: Vec<String>,
+        old: &'a [String],
+        new: &'a [String],
     },
```

Reply:
That would mean the lifetime would be added to `Job`, and store both the new and the old `Arc<Fingerprint>`s somewhere (e.g. in `Job`, so we run into the issue of lifetime-of-other-field). Alternatively, we could make the `Fingerprint::compare` method take in owned values and return them, skipping the need to clone, but then the `write_fingerprint` call wouldn't have a `Fingerprint` to work with. In the case of the first one, I guess it'd be possible and feasible to remove all the expensive-to-clone fields from `DirtyReason`, and to make the old and the new `Fingerprint` fields of `DirtyReasonPresentation`, to get them from there instead, but then the fingerprints themselves would live for a lot more than they probably should, and also the logs that use `<DirtyReason as Display>` would lose all those details.

Reply:
From the other angle, the message is quite useful, but only in verbose mode. I wonder if we can avoid the tiny overhead when not in verbose mode.

Reply:
It would be possible, maybe with another `enum DirtyReasonRef<'a>` and using the references from the `Fingerprint`s, coupled with some mechanism to turn a `DirtyReasonRef<'a>` to `DirtyReason` (maybe similar to `ToOwned`) by cloning the references, invoked only in verbose mode, but that would involve changing the `Result` of the `Fingerprint::compare` function to something like `Result<(), DirtyReasonRef<'a>>`, since `anyhow::Error`s have to be `'static`.

Reply:
You're totally right. It might not be that worthy, depending on how much code to change, as freshness check is not really in the hot path.